videostore/indexer.go

```diff
@@ -0,0 +1,512 @@
+}
+
+// videoRange represents a single contiguous block of stored video segments.
+type videoRange struct {
+	Start time.Time
+	End   time.Time
+}
+
+// videoRanges summarizes the state of the stored video segments.
+type videoRanges struct {
+	StorageUsedBytes int64
+	TotalDurationMs  int64
+	VideoCount       int
+	Ranges           []videoRange
+}
+
+// getVideoList returns the full list of video range structs from the db.
+func (ix *indexer) getVideoList(ctx context.Context) (videoRanges, error) {
+	if !ix.setupDone {
+		return videoRanges{}, errors.New("indexer setup not complete")
+	}
+
+	segments, err := ix.getSegmentsAscTime(ctx)
+	if err != nil {
+		return videoRanges{}, fmt.Errorf("failed to fetch segments for state: %w", err)
+	}
+
+	var videoRanges videoRanges
+	if len(segments) == 0 {
+		return videoRanges, nil
+	}
+
+	var prevRange *videoRange
+	for _, s := range segments {
+		videoRanges.VideoCount++
+		videoRanges.TotalDurationMs += s.DurationMs
+		videoRanges.StorageUsedBytes += s.SizeBytes
+
+		segmentStart := time.Unix(s.StartTimeUnix, 0)
+		segmentEnd := segmentStart.Add(time.Duration(s.DurationMs) * time.Millisecond)
+
+		if prevRange == nil {
+			prevRange = &videoRange{Start: segmentStart, End: segmentEnd}
+		} else {
+			if segmentStart.After(prevRange.End.Add(slopDuration)) {
+				// make a new range as there is too big of a gap between the prev segment and the new segment
+				videoRanges.Ranges = append(videoRanges.Ranges, *prevRange)
+				prevRange = &videoRange{Start: segmentStart, End: segmentEnd}
+			} else {
+				// extend range
+				prevRange.End = segmentEnd
+			}
+		}
+	}
+	if prevRange != nil {
+		videoRanges.Ranges = append(videoRanges.Ranges, *prevRange)
+	}
+
+	return videoRanges, nil
+}
+
+// getSegmentsAscTime is a helper function that retrieves all non-deleted segment data from the database,
+// ordered by start time.
+func (ix *indexer) getSegmentsAscTime(ctx context.Context) ([]segmentMetadata, error) {
+	query := fmt.Sprintf(`
+	SELECT file_name, start_time_unix, duration_ms, size_bytes
+	FROM %s
+	WHERE deleted_at IS NULL
+	ORDER BY start_time_unix ASC;
+	`, segmentsTableName)
+
+	rows, err := ix.db.QueryContext(ctx, query)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query all segments: %w", err)
+	}
+	defer rows.Close()
+
+	var segments []segmentMetadata
+
+	for rows.Next() {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+		var sm segmentMetadata
+
+		if err := rows.Scan(&sm.FileName, &sm.StartTimeUnix, &sm.DurationMs, &sm.SizeBytes); err != nil {
+			return nil, fmt.Errorf("failed to scan segment row during full query: %w", err)
+		}
+
+		segments = append(segments, sm)
+	}
+
+	if err := rows.Err(); err != nil {
+		return nil, fmt.Errorf("error iterating over all segments query result: %w", err)
+	}
+
+	ix.logger.Debugf("retrieved %d segments from index", len(segments))
+	return segments, nil
+}
+
+func (ix *indexer) close() error {
+	if ix.setupDone {
+		err := ix.db.Close()
```

Comment:
What happens if db has a method called on it after `Close()` is called?

If the answer is:
1. You get an error, but no crash / no undefined behavior, and assuming all code that calls db methods handles errors (without crashing etc) then making setupDone atomic should be sufficient
2. If it's undefined behavior / crashes / hangs / another unacceptable condition. We should wrap setupDone & db in a rw mutex

If 2:
1. Make a closedCtx on the struct
2. At beginning of close cancel that context
3. make a rw mutex on the struct that protects the db, write in close read everywhere elase
4. check closedCtx within the rw mutex / pass the decendent context to all the sqlite functions to see if close is trying to terminate the indexer

Reply:
```
2025-05-16T17:10:31.731Z	INFO	rdk:component:camera/video-store-1	videostore/indexer.go:526	DB: got db error after close: sql: database is closed
```

```
func (ix *indexer) close() error {
	if ix.setupDone.Load() {
		err := ix.db.Close()
		_, errDB := ix.db.Query("SELECT 1")
		if errDB != nil {
			ix.logger.Infof("DB: got db error after close: %v", errDB)
		} else {
			ix.logger.Error("DB: db connection still open after close")
		}
		if err != nil {
			return fmt.Errorf("error closing index db: %w", err)
		}
		ix.setupDone.Store(false)
		ix.logger.Debug("indexer closed")
	}
	return nil
}
```

looks like 1 is true